src/app/api/tasks/tasks.service.ts

```diff
@@ -0,0 +1,103 @@
+import { CreateTaskRequest, UpdateTaskRequest } from '@/types/dto/tasks.dto'
+import { BaseService } from '@api/core/services/base.service'
+import { Resource } from '@api/core/types/api'
+import { PoliciesService } from '@api/core/services/policies.service'
+import { AssigneeType } from '@prisma/client'
+
+type FilterByAssigneeId = {
+  assigneeId: string
+  assigneeType: AssigneeType
+}
+
+export class TasksService extends BaseService {
+  /**
+   * Builds filter for "get" service methods.
+   * If user is an IU, return filter for all tasks associated with this workspace
+   * If user is a client, return filter for just the tasks assigned to this clientId.
+   * If user is a client and has a companyId, return filter for just the tasks assigned to this clientId `OR` to this companyId
+   */
+  private buildReadFilters() {
+    const user = this.user
+
+    let filters = { where: { workspaceId: user.workspaceId, OR: undefined as unknown as FilterByAssigneeId[] } }
+
+    if (user.clientId) {
+      filters = {
+        where: {
+          ...filters.where,
+          OR: [{ assigneeId: user.clientId as string, assigneeType: 'client' }],
+        },
+      }
+    }
+    if (user.clientId && user.companyId) {
+      filters = {
+        where: {
+          ...filters.where,
+          OR: [
+            { assigneeId: user.clientId as string, assigneeType: 'client' },
+            { assigneeId: user.companyId, assigneeType: 'company' },
+          ],
+        },
+      }
+    }
+
+    return filters
+  }
+
+  async getAllTasks() {
+    const user = this.user
+    new PoliciesService(user).authorize('read', Resource.Tasks)
```

Comment:
lets not hard code read. we have enum for actions?

Reply:
How about

```
const policyGate = new PolicyGate();
policyGate.authorize(Action.READ, Resources.Task, user);
```
?

Reply:
@sazanrjb I did that initially but it felt like 'read' felt more natural to read since you can read it like
`authorize` `read` of `resource task` for `user`


Changed it to enum again.